python/cudf/cudf/_libxx/table.pyx

```diff
@@ -1,79 +1,144 @@
-from collections import OrderedDict
+
+import itertools
 
 import numpy as np
 
+from cython.operator cimport dereference
 from libc.stdint cimport uintptr_t
 
 from cudf._libxx.column cimport *
 from cudf._libxx.lib cimport *
+from cudf.utils.utils import OrderedColumnDict
 
 
-class OrderedColumnDict(OrderedDict):
-    def __setitem__(self, key, value):
-        from cudf.core.column import ColumnBase
+cdef class Table:
 
-        if not isinstance(value, ColumnBase):
-            raise TypeError(
-                f"Cannot insert object of type "
-                "{value.__class__.__name__} into OrderedColumnDict"
-            )
+    def __init__(self, data=None, index=None):
+        """
+        Data: an iterable of Columns
+        """
+        if data is None:
+            data = {}
+        self._data = OrderedColumnDict(data)
+        self._index = index
 
-        if self.first is not None and len(self.first) > 0:
-            if len(value) != len(self.first):
-                raise ValueError(
-                    f"Cannot insert Column of different length "
-                    "into OrderedColumnDict"
-                )
+    @property
+    def _column_names(self):
+        return tuple(self._data.keys())
 
-        super().__setitem__(key, value)
+    @property
+    def _index_names(self):
+        return tuple(self._index._data.keys())
 
     @property
-    def first(self):
+    def _columns(self):
         """
-        Returns the first value if self is non-empty;
-        returns None otherwise.
+        Return a list of Column objects backing this dataframe
         """
-        if len(self) == 0:
-            return None
-        else:
-            return next(iter(self.values()))
-
-
-cdef class _Table:
+        return tuple(self._data.values())
+    
+    @staticmethod
+    cdef Table from_unique_ptr(unique_ptr[table] c_tbl,
+                               column_names,
+                               index_names=None):
```

Comment:
May want to make this an empty tuple instead of `None` to simplify the logic where you can still length check and iterate over it.

Reply:
I think that `index_names=()` is different from `index_names=None` where the former might refer to an empty `Index`

Reply:
@shwina I think you need to handle `None` differently then, because if `index_names` is `None` you bypass the loop entirely instead of getting one index column with name `None`.

Reply:
Connected offline, `None` is used for when there's no index, if you want one index column with `None` as a name you need to pass it as a list-like `[None]`